libats/src/main/scala/com/advancedtelematic/libats/data/Urn.scala

```diff
@@ -0,0 +1,33 @@
+package com.advancedtelematic.libats.data
+
+import java.util.UUID
+import io.circe.{Decoder, Encoder}
+import scala.util.control.NoStackTrace
+
+object Urn {
+
+  abstract class Urn {
+    val nid: String
+    val nss: String
+    override def toString = s"urn:$nid:$nss"
+  }
+
+  case class IllegalUrnException(val msg: String) extends Exception(msg) with NoStackTrace
+
+  final case class CorrelationId(resource: String, uuid: UUID) extends Urn {
+    override val nid = CorrelationId.namespaceId
+    override val nss = s"$resource:${uuid.toString}"
+  }
+
+  object CorrelationId {
```

Comment:
What about 
```scala
sealed trait CorrelationId
final case class CampaignId(value: UUID) extends CorrelationId
final case class MultiTargetUpdate(value: UUID) extends CorrelationId

object CorrelationId {
  private[this] val CorrelationIdRe = """^urn:here-ota:(mtu|campaign):([0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12})$"""r

  implicit val DecoderInstance: Decoder[CorrelationId] = Decoder.decodeString.emap {
    case CorrelationIdRe("mtu", uuid) =>
      Right(MultiTargetUpdate(UUID.fromString(uuid)))

    case CorrelationIdRe("campaign", uuid) =>
      Right(CampaignId(UUID.fromString(uuid)))
    case x =>
      Left(s"'$x' is invalid correlationId")
  }

  implicit val EncoderInstance: Encoder[CorrelationId] = Encoder.encodeString.contramap {
    case CampaignId(x) => s"urn:here-ota:campaign:$x"
    case MultiTargetUpdate(x) => s"urn:here-ota:mtu:$x"
  }
}
```
I believe it gives us type safe correlation id and models the fact that urn is just a way it is encoded in json

Reply:
Cool!